### PR TITLE
mpl/atomic: Suppress Warnings with Solaris Studio

### PR DESCRIPTION
**Pull request by shintaro-iwasaki**
## Pull Request Description

This patch suppresses warnings in MPL/atomic when Solaris compilers are used.
The Solaris compiler shows the warnings on GCC builtins if arguments are const or volatile.
```c
  // gcc __atomic
  const int a = 0; // "int a" is okay with suncc
  int b = __atomic_load_n(&a, __ATOMIC_RELAXED);
  // gcc __sync
  volatile int c = 0; // "int c" is okay with suncc
  __sync_lock_test_and_set(&c, 1);
```
```
# suncc atomic.c // Oracle Developer Studio 12.6
"atomic.c", line *: warning: argument #2 is incompatible with prototype:
	prototype: pointer to void : "atomic.c", line 0
	argument : pointer to const int
"atomic.c", line *: warning: argument #2 is incompatible with prototype:
	prototype: pointer to void : "atomic.c", line 0
	argument : pointer to volatile int
```
Note that there is no problem with gcc/clang/icc.

This PR fixes it by adding a warning-suppression pragma as follows:
```c
#ifdef __SUNPRO_C
#pragma error_messages (off, E_ARG_INCOMPATIBLE_WITH_ARG_L)
#endif
// atomic.h
#ifdef __SUNPRO_C
#pragma error_messages (default, E_ARG_INCOMPATIBLE_WITH_ARG_L)
#endif
```

We can remove these pragmas if the Solaris compiler changes this behavior.

## Expected Impact

Warnings are suppressed with Solaris compilers. There is no effect with other compilers

## Author Checklist
* [ ] Reference appropriate issues (with "Fixes" or "See" as appropriate)
* [ ] Remove xfail from the test suite when fixing a test
* [ ] Commits are self-contained and do not do two things at once
* [ ] Commit message is of the form: `module: short description` and follows [good practice](https://chris.beams.io/posts/git-commit/)
* [ ] Passes whitespace checkers
* [ ] Passes warning tests
* [ ] Passes all tests
* [ ] Add comments such that someone without knowledge of the code could understand
* [ ] Add Devel Docs in the `doc/` directory for any new code design

### CODE REVIEW

**Comment by hzhou** [comment]
It'll be nice to narrow the scope, for example, putting `#pragma error_messages (on, ...)` and `#prama error_messages(off, ...)` only around the offending statements inside each inline functions, but of course I can see that is quite messy.

It is not ideal to turnoff warnings blindly. Leave it noisy may still be a better option.

**Comment by shintaro-iwasaki** [comment]
I agree that this is not the ideal solution. I just want to note ...
- This patch only disables warnings of `E_ARG_INCOMPATIBLE_WITH_ARG_L`.
- The scope of `error_message` is just 100 lines and does not include other headers.
- Other warnings/errors can be checked with other compilers (e.g., gcc, clang, icc).

If really really needed I can create an atomic wrapper only for Soalris Studio (maybe as well as PGI and some specific version of ICC #3997)

**Comment by hzhou** [comment]
>     * The scope of `error_message` is just 100 lines and does not include other headers.

I missed that you did turn the error_message back to default at the end. So good!

File: src/mpl/include/mpl_atomic_gcc_atomic.h [hunk 1]
```diff
@@ -9,6 +9,10 @@
 
 #include <stdint.h>
 
+#ifdef __SUNPRO_C
+#pragma error_messages (off, E_ARG_INCOMPATIBLE_WITH_ARG_L)
```
Comment (hzhou) [inline src/mpl/include/mpl_atomic_gcc_atomic.h]: Could you add a comment explaining the reason for this (briefly)?

File: src/mpl/include/mpl_atomic_gcc_sync.h [hunk 2]
```diff
@@ -9,6 +9,10 @@
 
 #include <stdint.h>
 
+#ifdef __SUNPRO_C
+#pragma error_messages (off, E_ARG_INCOMPATIBLE_WITH_ARG_L)
```
Comment (hzhou) [inline src/mpl/include/mpl_atomic_gcc_sync.h]: Add comment for this as well?

**Comment by hzhou** [comment]
@raffenet Are you ok with this solution?

**Comment by raffenet** [comment]
Looks OK to me.

**Pull request merged**